README.md
# Amount

[![Packagist Version](https://img.shields.io/packagist/v/byrokrat/amount.svg?style=flat-square)](https://packagist.org/packages/byrokrat/amount)
[![Build Status](https://img.shields.io/travis/byrokrat/amount/master.svg?style=flat-square)](https://travis-ci.org/byrokrat/amount)
[![Quality Score](https://img.shields.io/scrutinizer/g/byrokrat/amount.svg?style=flat-square)](https://scrutinizer-ci.com/g/byrokrat/amount)
[![Scrutinizer Coverage](https://img.shields.io/scrutinizer/coverage/g/byrokrat/amount.svg?style=flat-square)](https://scrutinizer-ci.com/g/byrokrat/amount/?branch=master)
[![Dependency Status](https://img.shields.io/gemnasium/byrokrat/amount.svg?style=flat-square)](https://gemnasium.com/byrokrat/amount)

Value objects for monetary amounts.

Features
--------
 * Immutable value object.
 * Using the [bcmath](http://php.net/manual/en/book.bc.php) extension for
   arbitrary floating point arithmetic precision.
 * [Currency](#working-with-currencies) support to prevent mixing currencies.
 * Out of the box support for ISO 4217 currencies.
 * Simple interface for [defining new currencies](#creating-new-currencies).
 * Support for multiple [rounding](#rounding) strategies.
 * Support for [allocating](#allocating) amounts based on ratios.
 * Support for the [signal string](#signal-strings) format as used in the swedish direct debit system.

Installation
------------
```shell
composer require byrokrat/amount:^2
```

Usage
-----
<!-- @expectOutput 1101.00 -->
```php
use byrokrat\amount\Amount;

$amount = new Amount('100.6');

// outputs 1 (true)
echo $amount->isGreaterThan(new Amount('50'));

// round to 0 decimal digits
$roundedAmount = $amount->roundTo(0);

// outputs 101.00
echo $roundedAmount;
```

API
---
[`Amount`](/src/Amount.php) defines the following api:

Method signature                                    | returns  | description
:-------------------------------------------------- | :------- | :------------------------------------------
**__construct(string $amount)**                     | Amount   | Create new instance
**getAmount()**                                     | string   | Get raw amount
**roundTo([int $precision, [Rounder $rounder]])**   | Amount   | Get new Amount rounded to $precision
**getString([int $precision, [Rounder $rounder]])** | string   | Get amount as string
**__tostring()**                                    | string   | Get amount as string
**getInt([Rounder $rounder])**                      | integer  | Get amount as integer (WARNING)
**getFloat([int $precision, [Rounder $rounder]])**  | float    | Get amount as float (WARNING)
**getSignalString([Rounder $rounder])**             | string   | Get amount as a signal string
**add(Amount $amount)**                             | Amount   | Get new Amount with $amount added
**subtract(Amount $amount)**                        | Amount   | Get new Amount with $amount subtracted
**multiplyWith(mixed $amount)**                     | Amount   | Get new Amount multiplied with $amount
**divideBy(mixed $amount)**                         | Amount   | Get new Amount divided by $amount
**compareTo(Amount $amount)**                       | integer  | 0 if equals, 1 if greater, -1 otherwise
**equals(Amount $amount)**                          | boolean  | Check if equals amount
**isLessThan(Amount $amount)**                      | boolean  | Check if less than amount
**isLessThanOrEquals(Amount $amount)**              | boolean  | Check if less than or equals amount
**isGreaterThan(Amount $amount)**                   | boolean  | Check if greater than amount
**isGreaterThanOrEquals(Amount $amount)**           | boolean  | Check if greater than or equals amount
**isZero()**                                        | boolean  | Check if amount is zero
**isPositive()**                                    | boolean  | Check if amount is greater than zero
**isNegative()**                                    | boolean  | Check if amount is less than zero
**getInverted()**                                   | Amount   | Get new amount with sign inverted
**getAbsolute()**                                   | Amount   | Get new amount with negative sign removed
**allocate(array $ratios, [int $precision])**       | Amount[] | Allocate amount based on list of ratios

Creating Amounts from other formats
-----------------------------------

### Floating point numbers

`Amount` contains two convenience methods for working with floating point numbers.
`createFromNumber` can create an Amount object from a floating point number,
`getFloat` can convert an Amount object to a float. These methods should be used
with care.

It is important to note that computers internally use the binary floating
point format and cannot accurately represent a number like 0.1, 0.2 or
0.3 at all. Using floating point numbers leads to a loss of precision.
For example `floor((0.1+0.7)*10)` will usually return 7 instead of the
expected 8, since the internal representation will be something like
7.9999999999999991118....

For this reason floats should never ne used to store monetary data. These
methods exists for rare situations when converting to or from native formats is
inevitable. Unless you know what you are doing they should **NOT** be used.

For more information see the [php manual](http://php.net/manual/en/language.types.float.php)
or [What Every Programmer Should Know About Floating-Point Arithmetic](http://floating-point-gui.de/).

### Formatted numbers

You can create Amounts from strings formatted with non-standard or locale dependent
decimal point and grouping characters using the static method **createFromFormat**.

<!-- @expectOutput 2000.50 -->
```php
use byrokrat\amount\Amount;

$formattedAmount = "2 000:50";

$amount = Amount::createFromFormat($formattedAmount, ":", " ");

echo $amount;  // outputs 2000.50
```

### Signal strings

The signal string format contans no decimal point and negative amounts are signaled
using a letter instead of the final digit. Create Amounts from signal strings
using the static method **createFromSignalString**.

Working with currencies
-----------------------
The currency subsystem helps prevent bugs where values in different currencies are
mixed (for example added together). Currency objects subclass `Amount` and works in
the same way, with the added feature that they know their defined currency.

<!-- @expectException Exception -->
```php
use byrokrat\amount\Currency\SEK;
use byrokrat\amount\Currency\EUR;

$sek = new SEK('100');

// throws an exception
$sek->add(new EUR('1'));
```

<!-- @expectOutput 101.00 -->
```php
use byrokrat\amount\Currency\SEK;

$sek = new SEK('100');

// works as intended, outputs 101.00
echo $sek->add(new SEK('1'));
```

### ISO 4217

A comprehensive set of ISO 4217 currencies are bundled in the
[`byrokrat\amount\Currency`](/src/Currency/) namespace.

Please note that the Turkish Lira with the ISO code `TRY` is represented as
`_TRY` as try is a reserved keyword.

### Creating new currencies

Creating new currencies however is straight forward. Simply subclass the
[`Currency`](/src/Currency.php) class and define `getCurrencyCode()`.

Additionaly you may override `getDisplayPrecision()`, `getInternalPrecision()` and
`getDefaultRounder()` to further define your currency's behaviour.

### Exchanging

Exchanging currencies is supported using `createFromExchange`. Note that you must
supply the correct exchange rate.

<!-- @expectOutput 9.27 -->
```php
use byrokrat\amount\Currency\SEK;
use byrokrat\amount\Currency\EUR;

// One euro is exchanged into swedish kronas using the exchange rate 9.27198929
// resulting in the value of SEK 9.27198929
echo $sek = SEK::createFromExchange(new EUR('1'), '9.27198929');
```

### Formatting currencies

Currency objects can easily be formatted using php's built in `NumberFormatter`.

<!-- @expectOutput /1 234 567[:,]89 €/ -->
```php
use byrokrat\amount\Currency\EUR;

// Create some amount of euros
$money = new EUR('1234567.89');

// Create a currency formatter with swedish formatting rules
$formatter = new NumberFormatter('sv_SE', NumberFormatter::CURRENCY);

// Format euros according to swedish standards, outputs 1 234 567:89 €
echo $formatter->formatCurrency($money->getFloat(), $money->getCurrencyCode());
```

Rounding
--------
A number of rounding strategies are supported. To implement your own see the
[Rounder](/src/Rounder.php) interface.

<!-- @expectOutput 2.001.001.002.002.001.001.002.002.001.00 -->
```php
namespace byrokrat\amount;
$amount = new Amount('1.5');

// outputs 2
echo $amount->roundTo(0, new Rounder\RoundUp);

// outputs 1
echo $amount->roundTo(0, new Rounder\RoundDown);

// outputs 1
echo $amount->roundTo(0, new Rounder\RoundTowardsZero);

// outputs 2
echo $amount->roundTo(0, new Rounder\RoundAwayFromZero);

// outputs 2
echo $amount->roundTo(0, new Rounder\RoundHalfUp);

// outputs 1
echo $amount->roundTo(0, new Rounder\RoundHalfDown);

// outputs 1
echo $amount->roundTo(0, new Rounder\RoundHalfTowardsZero);

// outputs 2
echo $amount->roundTo(0, new Rounder\RoundHalfAwayFromZero);

// outputs 2
echo $amount->roundTo(0, new Rounder\RoundHalfToEven);

// outputs 1
echo $amount->roundTo(0, new Rounder\RoundHalfToOdd);
```

For more info on rounding strategies see [wikipedia](https://en.wikipedia.org/wiki/Rounding).

Allocating
----------
Allocating is the process of dividing an amount based on ratios in such a way that
the smallest unit is not divided (every currency has a smallest unit that is
non-dividable) but instead handed to the receiver next in line.

The ratios can be seen as (but does not have to be) percentages. A hundred units
can thous be divided to two receivers as

<!-- @expectOutput 30.0070.00 -->
```php
use byrokrat\amount\Amount;
$money = new Amount('100');

list($receiverA, $receiverB) = $money->allocate([30, 70]);

// outputs 30
echo $receiverA;

// outputs 70
echo $receiverB;
```

The strength of allocating becomes clear when we distribute a value that we are
not able to divide evenly. In this case the order of the receivers is significant.

<!-- @expectOutput 0.020.030.040.01 -->
```php
use byrokrat\amount\Amount;

$money = new Amount('0.05');

list($receiverA, $receiverB) = $money->allocate([30, 70]);

// outputs 0.03
echo $receiverA;

// outputs 0.02
echo $receiverB;

list($receiverA, $receiverB) = $money->allocate([70, 30]);

// outputs 0.04
echo $receiverA;

// outputs 0.01
echo $receiverB;
```

In these examples the undividable unit used is `0.01`. This is the default behaviour.
Change it either by specifying the `$precision` parameter or by overriding
`getDisplayPrecision()` in your currency class.


CHANGES.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

- Require php 7 and strict typing.

## [1.0.2] - 2016-08-18

### Added
- Add a .gitattributes to prevent tests and unneeded files from being included in composer installs.

## [1.0.1] - 2016-05-04

### Added
- Integration testing of examples in README.md

## [1.0.0] - 2015-01-26
- Initial release


src/Amount.php
<?php

declare(strict_types = 1);

namespace byrokrat\amount;

/**
 * Value class for working with and representing monetary amounts
 *
 * Uses the bcmath extension for arbitrary floating point arithmetic precision
 */
class Amount
{
    /**
     * @var string
     */
    private $amount;

    /**
     * @var array Substitution map for signal strings
     */
    private static $signals = [
        '0'=>'å', '1'=>'J', '2'=>'K', '3'=>'L', '4'=>'M', '5'=>'N', '6'=>'O', '7'=>'P', '8'=>'Q', '9'=>'R'
    ];

    /**
     * Create amount from numerical string
     *
     * @throws InvalidArgumentException If $amount is not valid
     */
    public function __construct(string $amount)
    {
        if (!preg_match("/^[+-]?\d*\.?\d*$/", $amount)) {
            throw new InvalidArgumentException("Constructor expects a numerical string");
        }

        $this->amount = $amount;
    }

    /**
     * Create amount from integer or floating point number
     *
     * It is important to note that computers internally use the binary floating
     * point format and cannot accurately represent a number like 0.1, 0.2 or
     * 0.3 at all. Using floating point numbers leads to a loss of precision.
     * For example `floor((0.1+0.7)*10)` will usually return 7 instead of the
     * expected 8, since the internal representation will be something like
     * 7.9999999999999991118....
     *
     * For this reason floats should never ne used to store monetary data. This
     * method exists for rare situations when converting from native formats is
     * inevitable. Unless you know what you are doing it should NOT be used.
     *
     * @param  int|float $number
     * @throws InvalidArgumentException If $number is not an integer or float
     */
    public static function createFromNumber($number, int $precision = -1): Amount
    {
        if (!is_int($number) && !is_float($number)) {
            throw new InvalidArgumentException(
                "createFromNumber() expects an integer or float, found: " . gettype($number)
            );
        }

        return new static(
            number_format(
                $number,
                $precision >= 0 ? $precision : self::getInternalPrecision(),
                '.',
                ''
            )
        );
    }

    /**
     * Create amount from a formatted string
     */
    public static function createFromFormat(string $amount, string $point = '.', string $sep = ''): Amount
    {
        return new static(
            str_replace(
                [$point, $sep, '[~placeholder~]'],
                ['[~placeholder~]', '', '.'],
                $amount
            )
        );
    }

    /**
     * Create amount from signal string
     *
     * Signal strings does not contain a decimal digit separator. Instead the
     * last two digits are always considered decimals. For negative values the
     * last digit is converted to an alphabetic character according to schema:
     * 0 => å, 1 => J, 2 => K, ... 9 => R.
     *
     * @throws InvalidArgumentException If $signalStr is not a valid signal string
     */
    public static function createFromSignalString(string $signalStr): Amount
    {
        if (!preg_match("/^\d+(å|[JKLMNOPQR])?$/", $signalStr)) {
            throw new InvalidArgumentException("createFromSignalString() expects a valid signal string");
        }

        if (!is_numeric($signalStr)) {
            $signalStr = '-' . str_replace(
                self::$signals,
                array_keys(self::$signals),
                $signalStr
            );
        }

        return new static(
            preg_replace("/^(-?\d*)(\d\d)$/", "$1.$2", $signalStr, 1)
        );
    }

    /**
     * Get the raw stored amount
     */
    public function getAmount(): string
    {
        return $this->amount;
    }

    /**
     * Get new Amount rounded to $precision number of decimal digit using $rounder
     */
    public function roundTo(int $precision = -1, Rounder $rounder = null): Amount
    {
        return new static(
            ($rounder ?: $this->getDefaultRounder())->round(
                $this->getAmount(),
                $precision >= 0 ? $precision : $this->getDisplayPrecision()
            )
        );
    }

    /**
     * Get amount as string
     */
    public function getString(int $precision = -1, Rounder $rounder = null): string
    {
        return bcadd(
            $this->roundTo($precision, $rounder)->getAmount(),
            '0.0',
            $precision >= 0 ? $precision : $this->getDisplayPrecision()
        );
    }

    /**
     * Get amount as string
     */
    public function __tostring(): string
    {
        return $this->getString();
    }

    /**
     * Get amount as integer
     */
    public function getInt(Rounder $rounder = null): int
    {
        return (int)$this->getFloat(0, $rounder);
    }

    /**
     * Get amount as float
     *
     * It is important to note that computers internally use the binary floating
     * point format and cannot accurately represent a number like 0.1, 0.2 or
     * 0.3 at all. Using floating point numbers leads to a loss of precision.
     * For example `floor((0.1+0.7)*10)` will usually return 7 instead of the
     * expected 8, since the internal representation will be something like
     * 7.9999999999999991118....
     *
     * For this reason floats should never ne used to store monetary data. This
     * method exists for rare situations when converting to native formats is
     * inevitable. Unless you know what you are doing it should NOT be used.
     */
    public function getFloat(int $precision = -1, Rounder $rounder = null): float
    {
        return floatval($this->getString($precision, $rounder));
    }

    /**
     * Get amount as signal string
     *
     * Signal strings does not contain a decimal digit separator. Instead the
     * last two digits are always considered decimals. For negative values the
     * last digit is converted to an alphabetic character according to schema:
     * 0 => å, 1 => J, 2 => K, ... 9 => R.
     */
    public function getSignalString(Rounder $rounder = null): string
    {
        if ($this->isNegative()) {
            $signalStr = $this->getAbsolute()->getString(2, $rounder);
            $signalStr = substr($signalStr, 0, -1) . self::$signals[substr($signalStr, -1)];
        } else {
            $signalStr = $this->getString(2, $rounder);
        }

        return str_replace('.', '', $signalStr);
    }

    /**
     * Get new Amount with the value of $amount added to instance
     */
    public function add(Amount $amount, int $precision = -1): Amount
    {
        return new static(
            bcadd(
                $this->getAmount(),
                $amount->getAmount(),
                $precision >= 0 ? $precision : $this->getInternalPrecision()
            )
        );
    }

    /**
     * Get new Amount with the value of $amount subtracted from instance
     */
    public function subtract(Amount $amount, int $precision = -1): Amount
    {
        return new static(
            bcsub(
                $this->getAmount(),
                $amount->getAmount(),
                $precision >= 0 ? $precision : $this->getInternalPrecision()
            )
        );
    }

    /**
     * Get new Amount with the value of instance multiplied with $amount
     *
     * @param int|float|string|Amount $amount
     */
    public function multiplyWith($amount, int $precision = -1): Amount
    {
        return new static(
            bcmul(
                $this->getAmount(),
                $this->castToString($amount),
                $precision >= 0 ? $precision : $this->getInternalPrecision()
            )
        );
    }

    /**
     * Get new Amount with the value of instance divided by $amount
     *
     * @param int|float|string|Amount $amount
     */
    public function divideBy($amount, int $precision = -1): Amount
    {
        return new static(
            bcdiv(
                $this->getAmount(),
                $this->castToString($amount),
                $precision >= 0 ? $precision : $this->getInternalPrecision()
            )
        );
    }

    /**
     * Compare to amount
     *
     * @return int 0 if instance and $amount are equal, 1 if instance is larger, -1 otherwise.
     */
    public function compareTo(Amount $amount, int $precision = -1): int
    {
        return bccomp(
            $this->getAmount(),
            $amount->getAmount(),
            $precision >= 0 ? $precision : $this->getInternalPrecision()
        );
    }

    /**
     * Check if instance equals amount
     */
    public function equals(Amount $amount, int $precision = -1): bool
    {
        return 0 == $this->compareTo($amount, $precision);
    }

    /**
     * Check if instance is less than amount
     */
    public function isLessThan(Amount $amount, int $precision = -1): bool
    {
        return -1 == $this->compareTo($amount, $precision);
    }

    /**
     * Check if instance is less than or equals amount
     */
    public function isLessThanOrEquals(Amount $amount, int $precision = -1): bool
    {
        return $this->isLessThan($amount, $precision) || $this->equals($amount, $precision);
    }

    /**
     * Check if instance is greater than amount
     */
    public function isGreaterThan(Amount $amount, int $precision = -1): bool
    {
        return 1 == $this->compareTo($amount, $precision);
    }

    /**
     * Check if instance is greater than or equals amount
     */
    public function isGreaterThanOrEquals(Amount $amount, int $precision = -1): bool
    {
        return $this->isGreaterThan($amount, $precision) || $this->equals($amount, $precision);
    }

    /**
     * Check if amount is zero
     */
    public function isZero(int $precision = -1): bool
    {
        return $this->equals(new static('0'), $precision);
    }

    /**
     * Check if amount is greater than zero
     */
    public function isPositive(int $precision = -1): bool
    {
        return $this->isGreaterThan(new static('0'), $precision);
    }

    /**
     * Check if amount is less than zero
     */
    public function isNegative(int $precision = -1): bool
    {
        return $this->isLessThan(new static('0'), $precision);
    }

    /**
     * Get new Amount with sign inverted
     */
    public function getInverted(): Amount
    {
        return $this->isNegative() ? $this->getAbsolute() : new static('-' . str_replace('+', '', $this->getAmount()));
    }

    /**
     * Get new Amount with negative sign removed
     */
    public function getAbsolute(): Amount
    {
        return new static(str_replace('-', '', $this->getAmount()));
    }

    /**
     * Allocate amount based on a list of ratios
     *
     * @param  int[]|float[] $ratios List of ratios
     * @param  int $precision Optional decimal precision used in calculation
     * @return Amount[] The allocated amounts
     */
    public function allocate(array $ratios, int $precision = -1): array
    {
        $remainder = clone $this;
        $results = [];
        $total = array_sum($ratios);
        $precision = $precision >= 0 ? $precision : $this->getDisplayPrecision();
        $unit = new static(bcdiv('1', '1' . str_repeat('0', $precision), $precision));

        foreach ($ratios as $ratio) {
            $share = $this->multiplyWith($ratio)->divideBy($total)->roundTo($precision, new Rounder\RoundDown);
            $results[] = $share;
            $remainder = $remainder->subtract($share);
        }

        for ($i = 0; $remainder->isGreaterThanOrEquals($unit) > 0; $i++) {
            $results[$i] = $results[$i]->add($unit);
            $remainder = $remainder->subtract($unit);
        }

        return $results;
    }


    /**
     * Get the default display precision
     */
    public static function getDisplayPrecision(): int
    {
        return 2;
    }

    /**
     * Get the default internal precision used in computations
     */
    public static function getInternalPrecision(): int
    {
        return 10;
    }

    /**
     * Get default rounding strategy
     */
    public static function getDefaultRounder(): Rounder
    {
        return new Rounder\RoundHalfUp;
    }

    /**
     * Get a numerical string from input
     *
     * @param  int|float|string|Amount $amount
     * @throws InvalidArgumentException If $amount is does not evaluate to a numberical string
     */
    private function castToString($amount): string
    {
        switch (gettype($amount)) {
            case 'integer':
            case 'double':
                $amount = static::createFromNumber($amount);
                break;
            case 'string':
                $amount = new static($amount);
                break;
        }

        if (!$amount instanceof Amount) {
            throw new InvalidArgumentException("Supplied argument is not valid");
        }

        return $amount->getAmount();
    }
}
